subprojects/gst-plugins-good/gst/rtp/gstrtph266pay.c

```diff
@@ -351,9 +444,111 @@ _process_nalu (GstRtpH266Pay * rtph266pay, GstBuffer * nalu_buf)
   GST_DEBUG_OBJECT (rtph266pay, "NALU decoded: %" NALU_PTR_FORMAT,
       NALU_ARGS (nalu));
 
+  _update_xps_cache (rtph266pay, nalu);
   g_queue_push_tail (&rtph266pay->nalus, nalu);
 }
 
+static void
+_update_xps_cache (GstRtpH266Pay * rtph266pay, const Nalu * nalu)
+{
+  GHashTable *map = rtph266pay->xps_id_nalu_map;
+  guint8 type = nalu->type;
+  guint8 id = nalu->xps_id;
+  Nalu *new_nalu;
+  gpointer *key;
+
+  if (!NALU_IS_XPS (nalu))
+    return;
+
+  new_nalu = _nalu_copy (nalu);
+  new_nalu->au_start = FALSE;
+  new_nalu->au_end = FALSE;
+  key = GINT_TO_POINTER ((type << 8) | id);
+  if (g_hash_table_insert (map, key, new_nalu))
+    GST_DEBUG_OBJECT (rtph266pay, "XPS(%u,%u) cached", type, id);
+  else
+    GST_DEBUG_OBJECT (rtph266pay, "XPS(%u,%u) replaced", type, id);
+
+  rtph266pay->ts_last_xps_to_sent = _get_nalu_running_time (rtph266pay, nalu);
+}
+
+static void
+_clear_xps_cache (GstRtpH266Pay * rtph266pay)
+{
+  rtph266pay->ts_last_xps_to_sent = GST_CLOCK_TIME_NONE;
+  g_hash_table_remove_all (rtph266pay->xps_id_nalu_map);
+}
+
+static gboolean
+_can_insert_xps (GstRtpH266Pay * rtph266pay, const Nalu * nalu)
+{
+  if (rtph266pay->config_interval == 0)
+    return FALSE;               // Disabled
+  if (_nalu_is_rsv (nalu) || !NALU_IS_VCL (nalu))
+    return FALSE;               // Can't insert before this kind of NALU
+
+  GstClockTime ts_nalu = _get_nalu_running_time (rtph266pay, nalu);
+  GstClockTime ts_last_xps = rtph266pay->ts_last_xps_to_sent;
+  gboolean ts_last_xps_valid = GST_CLOCK_TIME_IS_VALID (ts_last_xps);
+  gboolean automatic_interval = rtph266pay->config_interval < 0;
+  gboolean xps_already_sent = ts_last_xps_valid && (ts_last_xps == ts_nalu);
+
+  if (automatic_interval) {
+    if (!NALU_IS_IDR (nalu) || xps_already_sent)
+      return FALSE;
+    GST_DEBUG_OBJECT (rtph266pay, "IDR detected: XPS can be sent");
+    return TRUE;
+  } else {
+    if (!ts_last_xps_valid)
```

Comment:
This condition should determine if _insert_xps_cache is called or not.

Reply:
I think it makes more sense as it is. Note that `ts_last_xps_to_sent` might change from call to call to `_can_insert_xps()`.